pkg/queue/protobuf_stats_reporter_test.go

```diff
@@ -0,0 +1,151 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"knative.dev/serving/pkg/autoscaler/metrics"
+	"knative.dev/serving/pkg/network"
+)
+
+var testStat = metrics.Stat{
+	PodName:                          "testPod",
+	AverageConcurrentRequests:        5.0,
+	AverageProxiedConcurrentRequests: 5.0,
+	ProxiedRequestCount:              100.0,
+	RequestCount:                     100.0,
+	ProcessUptime:                    20.0,
+}
+
+func TestReporterReport(t *testing.T) {
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			reporter := NewProtobufStatsReporter(pod, test.reportingPeriod)
+			// Make the value slightly more interesting, rather than microseconds.
+			reporter.startTime = reporter.startTime.Add(-5 * time.Second)
+			reporter.Report(network.RequestStatsReport{
+				AverageConcurrency:        test.concurrency,
+				AverageProxiedConcurrency: test.proxiedConcurrency,
+				RequestCount:              test.reqCount,
+				ProxiedRequestCount:       test.proxiedReqCount,
+			})
+			stat := reporter.stat.Load().(metrics.Stat)
+			if stat.RequestCount != test.expectedReqCount {
+				t.Errorf("stat.RequestCount = %v, want %v", stat.RequestCount, test.expectedReqCount)
+			}
+			if stat.AverageConcurrentRequests != test.expectedConcurrency {
+				t.Errorf("stat.AverageConcurrentRequests = %v, want %v", stat.AverageConcurrentRequests, test.expectedConcurrency)
+			}
+			if stat.ProxiedRequestCount != test.expectedProxiedRequestCount {
+				t.Errorf("stat.ProxiedRequestCount = %v, want %v", stat.ProxiedRequestCount, test.expectedProxiedRequestCount)
+			}
+			if stat.AverageProxiedConcurrentRequests != test.expectedProxiedConcurrency {
+				t.Errorf("stat.AverageProxiedConcurrentRequests = %v, want %v", stat.AverageProxiedConcurrentRequests, test.expectedProxiedConcurrency)
+			}
+			if got := stat.ProcessUptime; got < 5.0 || got > 6.0 {
+				t.Errorf("Got %v for process uptime, wanted 5.0 <= x < 6.0", got)
+			}
+		})
+	}
+}
+
+func TestProtoHandler(t *testing.T) {
+	metricsStat := atomic.Value{}
+	metricsStat.Store(testStat)
+
+	tests := []struct {
+		name     string
+		reporter ProtobufStatsReporter
+		errorMsg string
+	}{{
+		name:     "No metrics available",
+		reporter: ProtobufStatsReporter{},
+		errorMsg: "An error has occurred while serving metrics:\n\nno metrics available yet",
+	}, {
+		name: "Metrics available",
+		reporter: ProtobufStatsReporter{
+			reportingPeriod: time.Duration(1),
+			startTime:       time.Now(),
+			stat:            metricsStat,
+			podName:         "testPod"},
+	},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			req, err := http.NewRequest(http.MethodGet, "/metrics", nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			req.Header.Set("Accept-content", network.ProtoAcceptContent)
```

Comment:
What's this for? I don't think we need the header at all and `Accept-content` seems used nowhere else.

Reply:
I am trying to mimic the real http request which contains the header, it can be omitted since the tests just exercises the proto handler and the handler assumes that field has been tested earlier.

Reply:
Sure we can keep it, but please make it `Accept` then as all other parts of the system do too, I believe.